base/intfuncs.jl

```diff
@@ -195,6 +195,11 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
+# x^p for any literal integer p is lowered to x^Val{p},
+# to enable compile-time optimizations specialized to p.
+# However, we still need a fallback that calls the general ^:
+^{p}(x, ::Type{Val{p}}) = x^p
```

Comment:
I dunno, the more I think about it the more I think that the ambiguity warning is actually helpful here.  Particularly if we start inlining `x^Val{p}`, anyone specializing purely on the first argument (and not restricting `p::Number`) will want to decide what to do for the `Val{p}` case.